src/platforms/wayland/displayclient.cpp

```diff
@@ -0,0 +1,856 @@
+{
+}
+
+void mgw::DisplayClient::pointer_enter(
+    wl_pointer* /*pointer*/,
+    uint32_t /*serial*/,
+    wl_surface* surface,
+    wl_fixed_t /*x*/,
+    wl_fixed_t /*y*/)
+{
+    std::lock_guard<decltype(outputs_mutex)> lock{outputs_mutex};
+
+    for (auto const& out : bound_outputs)
+    {
+        if (surface == out.second->surface)
+        {
+            pointer_displacement = out.second->dcout.top_left - geometry::Point{};
+            break;
+        }
+    }
+}
+
+void mgw::DisplayClient::pointer_leave(wl_pointer* /*pointer*/, uint32_t /*serial*/, wl_surface* /*surface*/)
+{
+}
+
+void mgw::DisplayClient::pointer_motion(wl_pointer* pointer, uint32_t /*time*/, wl_fixed_t /*x*/, wl_fixed_t /*y*/)
+{
+    if (fake_pointer_frame)
+        pointer_frame(pointer);
+}
+
+void mgw::DisplayClient::pointer_button(
+    wl_pointer* pointer,
+    uint32_t /*serial*/,
+    uint32_t /*time*/,
+    uint32_t /*button*/,
+    uint32_t /*state*/)
+{
+    if (fake_pointer_frame)
+        pointer_frame(pointer);
+}
+
+void mgw::DisplayClient::pointer_axis(
+    wl_pointer* pointer,
+    uint32_t /*time*/,
+    uint32_t /*axis*/,
+    wl_fixed_t /*value*/)
+{
+    if (fake_pointer_frame)
+        pointer_frame(pointer);
+}
+
+void mgw::DisplayClient::pointer_frame(wl_pointer* /*pointer*/)
+{
+}
+
+void mgw::DisplayClient::pointer_axis_source(wl_pointer* /*pointer*/, uint32_t /*axis_source*/)
+{
+}
+
+void mgw::DisplayClient::pointer_axis_stop(wl_pointer* /*pointer*/, uint32_t /*time*/, uint32_t /*axis*/)
+{
+}
+
+void mgw::DisplayClient::pointer_axis_discrete(wl_pointer* /*pointer*/, uint32_t /*axis*/, int32_t /*discrete*/)
+{
+}
+
+void mgw::DisplayClient::touch_down(
+    wl_touch* /*touch*/,
+    uint32_t /*serial*/,
+    uint32_t /*time*/,
+    wl_surface* surface,
+    int32_t /*id*/,
+    wl_fixed_t /*x*/,
+    wl_fixed_t /*y*/)
+{
+    std::lock_guard<decltype(outputs_mutex)> lock{outputs_mutex};
+
+    for (auto const& out : bound_outputs)
+    {
+        if (surface == out.second->surface)
+        {
+            touch_displacement = out.second->dcout.top_left - geometry::Point{};
+            break;
+        }
+    }
+}
+
+void mgw::DisplayClient::touch_up(
+    wl_touch* /*touch*/,
+    uint32_t /*serial*/,
+    uint32_t /*time*/,
+    int32_t /*id*/)
+{
+}
+
+void mgw::DisplayClient::touch_motion(
+    wl_touch* /*touch*/,
+    uint32_t /*time*/,
+    int32_t /*id*/,
+    wl_fixed_t /*x*/,
+    wl_fixed_t /*y*/)
+{
+}
+
+void mgw::DisplayClient::touch_frame(wl_touch* /*touch*/)
+{
+}
+    
+void mgw::DisplayClient::touch_cancel(wl_touch* /*touch*/)
+{
+}
+
+void mgw::DisplayClient::touch_shape(
+    wl_touch* /*touch*/,
+    int32_t /*id*/,
+    wl_fixed_t /*major*/,
+    wl_fixed_t /*minor*/)
+{
+}
+
+void mgw::DisplayClient::touch_orientation(
+    wl_touch* /*touch*/,
+    int32_t /*id*/,
+    wl_fixed_t /*orientation*/)
+{
+}
+
+void mgw::DisplayClient::seat_capabilities(wl_seat* seat, uint32_t capabilities)
```

Comment:
This fails to handle being called multiple times; it will result in events being multiplied if it is called multiple times.

I don't know if that should be blocking (seat capabilities changing should be rare), but it should definitely have at least a TODO.

You should ideally be saving the `wl_pointer` etc across calls, and releasing them if the new capabilities do not contain the relevant capability.

Reply:
Adding TODO comment